Feature/validate url add

d88naimi opened the pull request:
Validating domain extension fix and added more description to the tooltip

Fixes this bug reported in manager-ui https://github.com/zesty-io/manager-ui/issues/924

![Screen Shot 2022-02-17 at 11 30 33 AM](https://user-images.githubusercontent.com/22800749/154557397-dc7f4b82-6fda-483c-895d-aeca65830bfa.png)

![Screen Shot 2022-02-17 at 11 34 18 AM](https://user-images.githubusercontent.com/22800749/154557430-19ffc282-2d35-4933-b6bf-43995f5fa09b.png)

d88naimi committed:
adding a valid domain extension check

d88naimi committed:
adding domain extension notes to tooltip

shrunyan commented on src/apps/properties/src/views/PropertyOverview/components/Domain/Domain.js:
```diff
@@ -37,12 +37,23 @@ export default class Domain extends Component {
         .replace(/http:\/\/|https:\/\//g, '')
-    }
 
+      if (!strippedDomain.slice(-4).includes('.')) {
```
TLD length can vary from 2 to N characters. We need to account for longer TLDs and if a subdomain is included.
example of a subdomain: www.example.org
example of a long tld: example.studio

shrunyan replied:
The rules for the input validation are as follows;

- the value MUST not contain a protocol. Those being either `http://` or `https://`
- the value MUST have a domain extension. The domain extension MUST be at least 2 characters and COULD be up to N characters
- the value MUST allow for subdomains
- the value MUST not have a path

Examples of **valid** values;
- example.org
- www.example.org
- subdmain.example.org
- multi.sub.domain.example.org
- example.co.uk

Examples of values which can be corrected to return a valid value;
- http://example.org
- https://example.org
- ftp://example.org
- example.org/path/to/page
- example.org/

Examples of invalid values which can not be corrected
- example
- example.i

_Solution should take an value and return either a corrected value or an error_

giseleblair replied:
Regarding  "the value MUST not have a path"
Why can't a subdirectory be configured as a domain on an instance? Are we (webengine) not equipped for it? Or why can't we have an instance configured this way?

shrunyan replied:
@giseleblair that's a good question. Does WebEngine now support sub-directory paths? I believe so, there are instances setup this way. Should we allow that to be a self serviceable feature? Does that require speaking with an account manager to get configured?

:point_up: maybe a @ardeay question.

giseleblair replied:
@shrunyan we do have a setup like this in production. Check out instance 8-82e6c3becf-wbm93d

d88naimi replied:
@shrunyan after speaking with @ardeay the path parts should NOT be altered and left to the user.

d88naimi committed:
Merge branch 'master' into feature/validate-url-add

d88naimi committed:
updated regex to remove www , http, https

d88naimi committed:
updated regex to validate urls

d88naimi committed:
remove log

d88naimi committed:
updating cypress test, added Validator cypress function in commands.js

d88naimi committed:
seperate file

d88naimi committed:
imported validator.js file

d88naimi committed:
removed clg

d88naimi commented:
@shrunyan  added validator url cypress test 

<img width="2330" alt="Screen Shot 2022-02-28 at 5 02 22 PM" src="https://user-images.githubusercontent.com/22800749/156085112-61099a22-54b1-4c45-b16b-232f9b6f70a4.png">

shrunyan commented on src/apps/properties/src/views/PropertyOverview/components/Domain/Validator.js:
```diff
@@ -0,0 +1,5 @@
+export const Validator = url => {
+  return /^[a-zA-Z0-9][a-zA-Z0-9-]{1,61}[a-zA-Z0-9](?:\.[a-zA-Z]{2,})+$/.test(
```
the validator should not return a boolean. it needs to return a corrected string or an error.

meaning if someone provides the value `https://example.org` we should not force them to correct this, we should do it for them as we know the intent but require the value to be `example.org`

shrunyan commented on cypress/integration/domain_validator.js:
```diff
@@ -0,0 +1,26 @@
+describe('Validate adding a URL', function() {
+  it('Check domain', () => {
+    // //PASS
+    assert.equal(Validator('www.example.com'), true)
```
See comment below. This should not be a boolean equality check.

d88naimi commented on src/apps/properties/src/views/PropertyOverview/components/Domain/Domain.js:
```diff
@@ -37,12 +38,25 @@ export default class Domain extends Component {
     if (this.state.domain) {
       strippedDomain = this.state.domain
         .toLowerCase()
         .replace(/http:\/\/|https:\/\//g, '')
```
@shrunyan  We are already providing corrective measures against https|http and then after the Validator.js check against TLDs.

d88naimi replied:
@shrunyan @agalin920 updated to cover all protocols

d88naimi committed:
now removing all protocols

d88naimi committed:
adding comments

d88naimi committed:
refactoring validator js file

d88naimi committed:
updated validator.js logic, added cypress test

d88naimi committed:
updating info tip

d88naimi commented:
Refactored the Validator function to be a url for formatter and TLD validator, updated all the cypress testing.

![Screen Shot 2022-04-20 at 8 46 19 AM](https://user-images.githubusercontent.com/22800749/164287371-15b203e7-b08b-48f0-8abf-802031b20eee.png)

shrunyan commented:
@agalin920 would you review this PR please
